README.md
# awesome_carousel

A Flutter plug-in aimed at providing various carousel widgets.

## Author

- **Author**: [Juncai Li](https://blog.csdn.net/qq_28550263?spm=1011.2415.3001.5343)
- **Email**: [291148484@163.com](291148484@163.com)
- **License**: [MIT](http://thispage.tech:9680/jclee1995/flutter-carousels/-/blob/master/LICENSE)
- Repository: [http://thispage.tech:9680/jclee1995/flutter-carousels](http://thispage.tech:9680/jclee1995/flutter-carousels.git)
- 中文文档：[http://thispage.tech:9680/jclee1995/flutter-carousels/-/blob/master/README_CN.md](http://thispage.tech:9680/jclee1995/flutter-carousels/-/blob/master/README_CN.md)

## Getting Started

Carousel component provided in this module is very easy to realize carousel function in various demand scenarios, and developers can customize carousel related effects and carousel progress indicators.

Here are two examples of using Carousel's default carousel progress indicator:

```
import 'package:awesome_carousel/carousels.dart';
import 'package:flutter/material.dart';

class WidgetCarouselDemo extends StatelessWidget {
  const WidgetCarouselDemo({Key? key});

  @override
  Widget build(BuildContext context) {
    // Get the screen width and height
    final double width = MediaQuery.of(context).size.width;
    final double height = MediaQuery.of(context).size.height;

    // Define a list of colors for carousel items
    final List<Color> colors = [
      Colors.amber,
      Colors.blue,
      Colors.red,
      Colors.green,
      const Color.fromARGB(255, 59, 255, 177),
    ];

    // Create a list of widget units for the carousel
    List<Widget> units = [];

    for (var i = 0; i < colors.length; i++) {
      units.add(
        Container(
          width: width,
          height: height,
          decoration: BoxDecoration(
            color: colors[i],
          ),
          child: Center(
            child: Text(
              'Page $i',
              style: const TextStyle(
                fontSize: 46,
                fontWeight: FontWeight.w900,
              ),
            ),
          ),
        ),
      );
    }

    return Column(
      children: [
        // Create a horizontal carousel with circular indicators
        Carousel(
          units,
          indicatorShape: BoxShape.circle,
          indicatorWidth: 16,
          currentIndicatorColor: Colors.purple,
        ),
        const Divider(),
        // Create a vertical carousel with rectangular indicators
        ClipRRect(
          borderRadius: BorderRadius.circular(40.0), // Rounded corner radius
          child: Carousel(
            units.reversed.toList(),
            scrollDirection: Axis.vertical,
            indicatorShape: BoxShape.rectangle,
            indicatorWidth: 10,
            indicatorHeight: 10,
            currentIndicatorColor: Colors.pink,
            onUnitTapped: (index) => print('Carousel unit $index was tapped'),
            onIndicatorTapped: (index) => print('Progress indicator unit $index was tapped'),
          ),
        )
      ],
    );
  }
}
```

The effect is as follows:

![1697452585612](./image/README_CN/1697452585612.gif)

In order to achieve more complex functions, it is necessary to define more parameters and use the carousel controller. The following is an example of customizing the carousel progress indicator style and using network pictures as carousel units.

```dart
import 'package:flutter/material.dart';
import 'package:awesome_carousel/carousels.dart';

class NetworkImageCarouselDemo extends StatelessWidget {
  final List<String> imageUrls;

  final CarouselController _controller;

  const NetworkImageCarouselDemo(
    this.imageUrls,
    this._controller, {
    super.key,
  });

  @override
  Widget build(BuildContext context) {
    return Column(
      children: [
        Carousel(
          _buildImages(context),
          height: 400.0, // 设置轮播图高度
          controller: _controller,
          indicatorColor: const Color.fromARGB(255, 190, 255, 130),
          // currentIndicatorColor: Colors.transparent,
          indicatorBuilder: _indicatorBuilder,
          // 当用户点击图像时触发的回调函数
          onUnitTapped: (int index) {
            print('点击了第 $index 张图片');
          },
        ),
        Row(
          mainAxisAlignment: MainAxisAlignment.center,
          children: [
            ElevatedButton(
              onPressed: () {
                _controller.toFirstPage();
              },
              child: const Text('首页'),
            ),
            const Spacer(
              flex: 1,
            ),
            ElevatedButton(
              onPressed: () {
                _controller.toPreviousPage();
              },
              child: const Text('上一页'),
            ),
            const Spacer(
              flex: 1,
            ),
            ElevatedButton(
              onPressed: () {
                _controller.toNextPage();
              },
              child: const Text('下一页'),
            ),
            const Spacer(
              flex: 1,
            ),
            ElevatedButton(
              onPressed: () {
                _controller.toLastPage();
              },
              child: const Text('尾页'),
            ),
          ],
        )
      ],
    );
  }

  List<Widget> _buildImages(BuildContext context) {
    List<Widget> res = [];
    for (String url in imageUrls) {
      res.add(
        Image.network(
          url,
          fit: BoxFit.cover, // 使图片宽度占满
          width: MediaQuery.of(context).size.width, // 设置宽度为屏幕宽度
          errorBuilder: (context, error, stackTrace) {
            // 在加载失败时返回一个占位图，确保图片占用固定高度
            return Container(
              height: MediaQuery.of(context).size.height,
              color: Colors.grey,
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  const Text(
                    'Image loading error:',
                    style: TextStyle(
                      color: Colors.red,
                    ),
                  ),
                  Text(
                    "原因: $error\n追溯：$stackTrace",
                    style: const TextStyle(
                      color: Colors.white,
                    ),
                    maxLines: 9,
                    overflow: TextOverflow.ellipsis,
                  ),
                ],
              ),
            );
          },
        ),
      );
    }
    return res;
  }

  Map<String, Widget> _indicatorBuilder(int index) {
    return {
      'plain': Container(
        padding: const EdgeInsets.all(10),
        child: Column(
          children: [
            const Text(
              'plain',
              style: TextStyle(
                color: Colors.grey,
              ),
            ),
            Container(
              width: 30,
              height: 10,
              decoration: const BoxDecoration(color: Colors.white),
            ),
          ],
        ),
      ),
      'active': Container(
        padding: const EdgeInsets.all(10),
        child: Column(
          children: [
            const Icon(
              Icons.arrow_circle_down,
              color: Colors.amber,
            ),
            const Text(
              'active',
              style: TextStyle(
                color: Colors.red,
                fontSize: 20,
                fontWeight: FontWeight.w700,
              ),
            ),
            Container(
              width: 30,
              height: 10,
              decoration: BoxDecoration(
                color: switch (index) {
                  0 => Colors.blue,
                  1 => Colors.red,
                  2 => Colors.green,
                  3 => Colors.pink,
                  _ => Colors.purple,
                },
              ),
            ),
          ],
        ),
      )
    };
  }
}

```

The effect is as follows:

![1697453153655](./image/README_CN/1697453153655.gif)

As you can see, because you can completely customize the carousel indicator and carousel items, as long as you have enough imagination, you can package some complex but exquisite carousels. For example, in the e-commerce scene, the first carousel unit is the introduction video of the goods, and the corresponding carousel indicator is the progress bar of the video, while the subsequent carousel unit is only the carousel picture. For example:

![1697454139296](./image/README_CN/1697454139296.jpg)

This can be easily realized by integrating the corresponding video player module with this module.

### Network Carousel

> Added in v1.0.0+1.  

A network carousel is a widely used form of carousel, especially in today's web applications, that displays a set of images in a cyclic manner. It is derived from the generic carousel class, `Carousel`, and is represented by the `NetImgCarousel` class. With `NetImgCarousel`, you can load a collection of images from URLs for a network carousel and customize various parameters related to images and the network.

Here is an example of how to use the `NetImgCarousel` class.

```dart
import 'package:flutter/material.dart';
import 'package:awesome_carousel/carousels.dart' show NetImgCarousel;

class NetImgCarouselExample extends StatelessWidget {
  final List<String> imageUrls = [
    'https://gitee.com/jacklee1995/example-pictures/raw/master/shoe/shoe1(1).png',
    'https://gitee.com/jacklee1995/example-pictures/raw/master/shoe/shoe1(3).png',
    'https://gitee.com/jacklee1995/example-pictures/raw/master/shoe/shoe1(5).png',
    'https://gitee.com/jacklee1995/example-pictures/raw/master/shoe/shoe1(7).png',
    'https://gitee.com/jacklee1995/example-pictures/raw/master/shoe/shoe1(2).png',
  ];

  NetImgCarouselExample({super.key});

  @override
  Widget build(BuildContext context) {
    return Center(
      child: NetImgCarousel(
        imageUrls,
        height: 400.0,
        indicatorColor: Colors.grey,
        currentIndicatorColor: Colors.blue,
        onUnitTapped: (index) {
          // Carousel unit callback event
          print('Tapped on unit $index');
        },
        onIndicatorTapped: (index) {
          // Carousel indicator callback event
          print('Tapped on indicator $index');
        },
      ),
    );
  }
}
```

The running result of the code is as follows:

![Alt text](./WeChat_IxjFbKajq3.gif)

The `NetImgCarousel` class allows you to create network carousels with ease, making it convenient to display a collection of images from web URLs.

The following table describes the parameters of the `netimgroom` constructor:

| Parameter Name                     | Type                        | Default Value          | Description                                                                             |
| ---------------------------------- | --------------------------- | ---------------------- | --------------------------------------------------------------------------------------- |
| `imageUrls`                        | `List<String>`              |                        | List of network image URLs to be displayed in the carousel.                             |
| `height`                           | `double`                    | `400.0`                | Height of the carousel component.                                                       |
| `width`                            | `double`                    | `0.0`                  | Width of the carousel component.                                                        |
| `useindicator`                     | `bool`                      | `true`                 | Whether to use indicators, defaults to `true`.                                          |
| `onUnitTapped`                     | `FunctionWithAInt?`         |                        | Callback function when a unit in the carousel is clicked, optional.                     |
| `onIndicatorTapped`                | `FunctionWithAInt?`         |                        | Callback function when an indicator is clicked, optional.                               |
| `indicatorBuilder`                 | `FunctionIndicatorBuilder?` |                        | Custom indicator builder function, optional.                                            |
| `indicatorColor`                   | `Color`                     | `Colors.white`         | Default indicator color.                                                                |
| `currentIndicatorColor`            | `Color`                     | `Colors.blue`          | Color of the currently selected indicator.                                              |
| `indicatorWidth`                   | `double`                    | `40.0`                 | Indicator width.                                                                        |
| `indicatorHeight`                  | `double`                    | `26.0`                 | Indicator height.                                                                       |
| `indicatorMargin`                  | `double`                    | `3.0`                  | Spacing between indicators.                                                             |
| `indicatorToBottom`                | `double`                    | `10.0`                 | Distance of indicators from the bottom.                                                 |
| `indicatorShape`                   | `BoxShape`                  | `BoxShape.rectangle`   | Default indicator shape.                                                                |
| `pageSnapping`                     | `bool`                      | `true`                 | Whether to enable page snapping, defaults to `true`.                                    |
| `padEnds`                          | `bool`                      | `true`                 | Whether to add extra pages at the beginning and end of the carousel, default is `true`. |
| `clipBehavior`                     | `Clip`                      | `Clip.hardEdge`        | The behavior for the clip.                                                              |
| `reverse`                          | `bool`                      | `false`                | Whether to reverse the order of items.                                                  |
| `scrollDirection`                  | `Axis`                      | `Axis.horizontal`      | The axis along which the page view scrolls.                                             |
| `controller`                       | `CarouselController?`       |                        | Carousel controller, optional.                                                          |
| `disableIndicatorDefaultCallbacks` | `bool`                      | `false`                | Whether to disable default indicator callback functions.                                |
| `imgScale`                         | `double`                    | `1.0`                  | Image scale factor, default is `1.0`.                                                   |
| `imgFrameBuilder`                  | `ImageFrameBuilder?`        |                        | Image frame builder, optional.                                                          |
| `imgLoadingBuilder`                | `ImageLoadingBuilder?`      |                        | Image loading builder, optional.                                                        |
| `imgErrorBuilder`                  | `ImageErrorWidgetBuilder?`  |                        | Image error widget builder, optional.                                                   |
| `imgExcludeFromSemantics`          | `bool`                      | `false`                | Whether to exclude the image from semantics, default is `false`.                        |
| `imgSemanticLabel`                 | `String?`                   |                        | Image semantic label, optional.                                                         |
| `imgWidth`                         | `double?`                   |                        | Image width, optional.                                                                  |
| `imgHeight`                        | `double?`                   |                        | Image height, optional.                                                                 |
| `imgColor`                         | `Color?`                    |                        | Image color, optional.                                                                  |
| `imgOpacity`                       | `Animation<double>?`        |                        | Image opacity, optional.                                                                |
| `imgColorBlendMode`                | `BlendMode?`                |                        | Image color blend mode, optional.                                                       |
| `imgFit`                           | `BoxFit?`                   |                        | Image fit, optional.                                                                    |
| `imgCenterSlice`                   | `Rect?`                     |                        | Image center slice, optional.                                                           |
| `imgAlignment`                     | `AlignmentGeometry`         | `Alignment.center`     | Image alignment, optional.                                                              |
| `imgRepeat`                        | `ImageRepeat`               | `ImageRepeat.noRepeat` | Image repeat mode, optional.                                                            |
| `imgMatchTextDirection`            | `bool`                      | `false`                | Whether the image should match the text direction, optional.                            |
| `imgGaplessPlayback`               | `bool`                      | `false`                | Whether to enable gapless playback for the image, optional.                             |
| `imgFilterQuality`                 | `FilterQuality`             | `FilterQuality.low`    | Image filter quality, optional.                                                         |
| `imgIsAntiAlias`                   | `bool`                      | `false`                | Whether the image should be anti-aliased, optional.                                     |
| `imgHeaders`                       | `Map<String, String>?`      |                        | Headers for the network image, optional.                                                |
| `imgCacheWidth`                    | `int?`                      |                        | Image cache width, optional.                                                            |
| `imgCacheHeight`                   | `int?`                      |                        | Image cache height, optional.                                                           |

### Local Image Carousel

> Added in v1.0.0+1. 

Local resource picture carousel is also a very common type of carousel. In Flutter, you need to configure the `assets` option of the `pubspec.yaml` file first to ensure that your pictures can be loaded normally.

Here's a simple example:

```dart
import 'package:flutter/material.dart';
import 'package:awesome_carousel/img_carousels/assert_img_carousel.dart';

class AssetImgCarouselExample extends StatelessWidget {
  final List<String> assetImagePaths = [
    'assets/image1.png',
    'assets/image2.png',
    'assets/image3.png',
  ];

  AssetImgCarouselExample({super.key});

  @override
  Widget build(BuildContext context) {
    return Center(
      child: AssetImgCarousel(
        assetImagePaths,
        height: 400.0,
        indicatorColor: Colors.grey,
        currentIndicatorColor: Colors.blue,
        indicatorShape: BoxShape.circle,
        onUnitTapped: (index) {
          print('Tapped on unit $index');
        },
        onIndicatorTapped: (index) {
          print('Tapped on indicator $index');
        },
      ),
    );
  }
}
```

The code produces the following result:

![Alt text](./WeChat_wLFeQijQol.gif)

The table below describes the parameters of the `AssetImgCarousel` constructor:

| Parameter Name                   | Parameter Type            | Description                                                                           |
| -------------------------------- | ------------------------- | ------------------------------------------------------------------------------------- |
| assetImagePaths                  | List<String>              | List of asset image file paths for the carousel.                                      |
| height                           | double                    | Height of the carousel component.                                                     |
| width                            | double                    | Width of the carousel component.                                                      |
| useindicator                     | bool                      | Whether to use indicators, defaults to true.                                          |
| onUnitTapped                     | FunctionWithAInt?         | Callback function when a unit in the carousel is clicked, optional.                   |
| onIndicatorTapped                | FunctionWithAInt?         | Callback function when an indicator is clicked, optional.                             |
| indicatorBuilder                 | FunctionIndicatorBuilder? | Custom indicator builder function, optional.                                          |
| indicatorColor                   | Color                     | Default indicator color.                                                              |
| currentIndicatorColor            | Color                     | Color of the currently selected indicator.                                            |
| indicatorWidth                   | double                    | Indicator width.                                                                      |
| indicatorHeight                  | double                    | Indicator height.                                                                     |
| indicatorMargin                  | double                    | Spacing between indicators.                                                           |
| indicatorToBottom                | double                    | Distance of indicators from the bottom.                                               |
| indicatorShape                   | BoxShape                  | Default indicator shape, defaults to a rectangle.                                     |
| pageSnapping                     | bool                      | Whether to enable page snapping, defaults to true.                                    |
| padEnds                          | bool                      | Whether to add extra pages at the beginning and end of the carousel, default is true. |
| clipBehavior                     | Clip                      | The behavior for the clip.                                                            |
| reverse                          | bool                      | Whether to reverse the order of items.                                                |
| scrollDirection                  | Axis                      | The axis along which the page view scrolls.                                           |
| controller                       | CarouselController?       | Carousel controller, optional.                                                        |
| disableIndicatorDefaultCallbacks | bool                      | Whether to disable default indicator callback functions.                              |
| imgScale                         | double                    | Image scale factor.                                                                   |
| imgFrameBuilder                  | ImageFrameBuilder?        | Custom image frame builder, optional.                                                 |
| imgExcludeFromSemantics          | bool                      | Whether the image should be excluded from semantics, default is false.                |
| imgAlignment                     | AlignmentGeometry         | Alignment of the image within its frame.                                              |
| imgRepeat                        | ImageRepeat               | How the image should repeat.                                                          |
| imgMatchTextDirection            | bool                      | Whether the image should match the text direction.                                    |
| imgGaplessPlayback               | bool                      | Whether to enable gapless playback of images, default is false.                       |
| imgFilterQuality                 | FilterQuality             | The filter quality for the image, default is FilterQuality.low.                       |
| imgIsAntiAlias                   | bool                      | Whether the image should be anti-aliased.                                             |
| imgErrorBuilder                  | ImageErrorWidgetBuilder?  | Custom image error widget builder, optional.                                          |
| imgSemanticLabel                 | String?                   | Semantic label for the image, optional.                                               |
| imgWidth                         | double?                   | Image width, optional.                                                                |
| imgHeight                        | double?                   | Image height, optional.                                                               |
| imgColor                         | Color?                    | Color applied to the image, optional.                                                 |

CHANGELOG.md
# Changelog

## [1.0.0+2] - 2023-10-17

### Added

Introduce network image carousel widget NetImgCarousel class.

Introduce assets image carousel widget AssetCarousel class.

## [1.0.0+1] - 2023-10-16

### Added

Introduce general carousel widget Carousel class.


lib/base/base.dart
/// Carousel 是一个用于在Flutter应用程序中实现轮播的插件
///
/// 作者：李俊才
/// 邮箱：291148484@163.com
/// 协议：MIT
///
/// 有关MIT协议的详细信息，请参阅项目的许可证文档（通常在项目的根目录中的"LICENSE"文件）
///

import 'package:flutter/material.dart';
import '../types.dart';
import 'controller.dart';

/// Used to create a generic carousel component that can be customized for various carousels.
///
/// 用于创建通用的轮播组件，可为各种轮播自定义
///
/// - [units] represents a list of components to be carousel-ed.
/// - [units] 表示要轮播的组件列表
///
/// - [height] represents the height of the carousel component, defaults to 400.0.
/// - [height] 表示轮播组件的高度，默认为400.0
///
/// - [width] represents the width of the carousel component, defaults to 0.0.
/// - [width] 表示轮播组件的宽度，默认为0.0
///
/// - [onUnitTapped] represents a callback function when a carousel unit is tapped, optional.
/// - [onUnitTapped] 表示当轮播单元被点击时的回调函数，可选
///
/// - [onIndicatorTapped] represents a callback function when an indicator is tapped, optional.
/// - [onIndicatorTapped] 表示当指示器被点击时的回调函数，可选
///
/// - [useindicator] indicates whether to use indicators, defaults to true.
/// - [useindicator] 表示是否使用指示器，默认为true
///
/// - [indicatorColor] represents the default indicator color, defaults to white.
/// - [indicatorColor] 表示默认指示器的颜色，默认为白色
///
/// - [currentIndicatorColor] represents the color of the currently selected indicator, defaults to blue.
/// - [currentIndicatorColor] 表示当前选中指示器的颜色，默认为蓝色
///
/// - [indicatorWidth] represents the width of the indicators, defaults to 40.0.
/// - [indicatorWidth] 表示指示器的宽度，默认为40.0
///
/// - [indicatorHeight] represents the height of the indicators, defaults to 26.0.
/// - [indicatorHeight] 表示指示器的高度，默认为26.0
///
/// - [indicatorMargin] represents the spacing between the indicators, defaults to 3.0.
/// - [indicatorMargin] 表示指示器之间的间距，默认为3.0
///
/// - [indicatorToBottom] represents the distance of the indicators from the bottom, defaults to 10.0.
/// - [indicatorToBottom] 表示指示器距离底部的距离，默认为10.0
///
/// - [indicatorShape] represents the default shape of the indicators, defaults to a rectangle.
/// - [indicatorShape] 表示默认指示器的形状，默认为矩形
///
/// - [pageSnapping] indicates whether to enable page snapping, defaults to true.
/// - [pageSnapping] 表示是否启用页面快速吸附，默认为true
///
/// - [padEnds] indicates whether to add extra pages at the beginning and end of the carousel, default is true.
/// - [padEnds] 表示是否在轮播首尾添加额外的页面，默认为true
///
/// - [controller] represents the carousel controller, optional.
/// - [controller] 表示轮播控制器，可选
///
/// - [disableIndicatorDefaultCallbacks] Whether to disable default indicator callback functions.
/// - [disableIndicatorDefaultCallbacks] 是否禁用默认的指示器回调函数
///
class Carousel extends StatefulWidget {
  /// List of components to be carousel-ed.
  /// 要轮播的组件列表
  final List<Widget> units;

  /// Height of the carousel component.
  /// 轮播组件的高度
  final double height;

  /// Width of the carousel component.
  /// 轮播组件的宽度
  final double width;

  /// Whether to use indicators, defaults to true.
  /// 是否使用指示器，默认为true
  final bool useindicator;

  /// Callback function when a unit in the carousel is clicked, optional.
  /// 当轮播单元被点击时的回调函数，可选
  final FunctionWithAInt? onUnitTapped;

  /// Callback function when an indicator is clicked, optional.
  /// 当指示器被点击时的回调函数，可选
  final FunctionWithAInt? onIndicatorTapped;

  /// Custom indicator builder function, optional.
  /// 自定义指示器构建函数，可选
  final FunctionIndicatorBuilder? indicatorBuilder;

  /// Default indicator color.
  /// 默认指示器颜色
  final Color indicatorColor;

  /// Color of the currently selected indicator.
  /// 当前选中指示器的颜色
  final Color currentIndicatorColor;

  /// Indicator width.
  /// 指示器宽度
  final double indicatorWidth;

  /// Indicator height.
  /// 指示器高度
  final double indicatorHeight;

  /// Spacing between indicators.
  /// 指示器之间的间距
  final double indicatorMargin;

  /// Distance of indicators from the bottom.
  /// 指示器距离底部的距离
  final double indicatorToBottom;

  /// Default indicator shape, defaults to a rectangle.
  /// 默认指示器形状，默认为矩形
  final BoxShape indicatorShape;

  /// Whether to enable page snapping, defaults to true.
  /// 是否启用页面快速吸附，默认为true
  final bool pageSnapping;

  /// Whether to add extra pages at the beginning and end of the carousel, default is true.
  /// 是否在轮播首尾添加额外的页面，默认为true
  final bool padEnds;

  /// The behavior for the clip.
  /// 剪裁的行为
  final Clip clipBehavior;

  /// Whether to reverse the order of items.
  /// 是否倒序项目的顺序
  final bool reverse;

  /// The axis along which the page view scrolls.
  /// 页面视图滚动的轴
  final Axis scrollDirection;

  /// Carousel controller, optional.
  /// 轮播控制器，可选
  final CarouselController? controller;

  /// Whether to disable default indicator callback functions, default is false.
  /// 是否禁用默认指示器回调函数，默认为false。
  final bool disableIndicatorDefaultCallbacks;

  const Carousel(
    this.units, {
    super.key,
    this.height = 400.0,
    this.width = 0.0,
    this.onUnitTapped,
    this.onIndicatorTapped,
    this.useindicator = true,
    this.indicatorColor = Colors.white,
    this.currentIndicatorColor = Colors.blue,
    this.indicatorWidth = 40.0,
    this.indicatorHeight = 26.0,
    this.indicatorMargin = 3.0,
    this.indicatorToBottom = 10.0,
    this.indicatorShape = BoxShape.rectangle,
    this.indicatorBuilder,
    this.pageSnapping = true,
    this.padEnds = true,
    this.clipBehavior = Clip.hardEdge,
    this.reverse = false,
    this.scrollDirection = Axis.horizontal,
    this.disableIndicatorDefaultCallbacks = false,
    this.controller,
  });

  @override
  State<Carousel> createState() => _CarouselState();
}

class _CarouselState extends State<Carousel> {
  double _width = 0.0;
  List<Map<String, Widget>> customIndicators = [];
  late CarouselController controller;

  @override
  Widget build(BuildContext context) {
    controller = widget.controller ??
        CarouselController(
          widget.units.length,
          initialPage: 0,
        );
    if (widget.width <= 0) {
      _width = MediaQuery.of(context).size.width;
    } else {
      _width = widget.width;
    }

    if (widget.indicatorBuilder != null) {
      for (int index = 0; index < widget.units.length; index++) {
        customIndicators.add(
          widget.indicatorBuilder!(index),
        );
      }
    }

    return Column(
      children: <Widget>[
        Stack(
          children: [
            SizedBox(
              height: widget.height,
              child: ListenableBuilder(
                listenable: controller,
                builder: (BuildContext context, Widget? child) {
                  return PageView.builder(
                    scrollDirection: widget.scrollDirection,
                    controller: controller.pageController,
                    itemCount: widget.units.length,
                    pageSnapping: widget.pageSnapping,
                    padEnds: widget.padEnds,
                    clipBehavior: widget.clipBehavior,
                    reverse: widget.reverse,
                    onPageChanged: (int page) {
                      controller.currentPage = page;
                    },
                    itemBuilder: (context, index) {
                      return GestureDetector(
                        // 调用传入的点击事件
                        onTap: () {
                          widget.onUnitTapped!(index);
                        },
                        child: Align(
                          alignment: Alignment.bottomCenter,
                          child: widget.units[index],
                        ),
                      );
                    },
                  );
                },
              ),
            ),
            widget.useindicator
                ? Positioned(
                    bottom: widget.indicatorToBottom,
                    child: SizedBox(
                      width: _width,
                      child: Align(
                        alignment: Alignment.bottomCenter,
                        child: ListenableBuilder(
                          listenable: controller,
                          builder: (BuildContext context, Widget? child) {
                            return _buildProgressIndicator();
                          },
                        ),
                      ),
                    ),
                  )
                : Container(),
          ],
        ),
      ],
    );
  }

  /// 构建进度指示器
  Widget _buildProgressIndicator() {
    final List<Widget> indicators = [];

    for (int index = 0; index < widget.units.length; index++) {
      indicators.add(
        GestureDetector(
          // 当前进度指示器事件，允许禁用默认事件和指定自定义事件
          onTap: () {
            // 未禁用默认事件时，点击后跳转页面
            if (!widget.disableIndicatorDefaultCallbacks) {
              controller.goToPage(index);
            }
            // 若指定用户自定义事件则调用之
            if (widget.onIndicatorTapped != null) {
              widget.onIndicatorTapped!(index);
            }
          },
          // 但用户指示器列表非空时，按照
          child: customIndicators.isEmpty
              ? Container(
                  width: widget.indicatorWidth,
                  height: widget.indicatorHeight,
                  margin:
                      EdgeInsets.symmetric(horizontal: widget.indicatorMargin),
                  decoration: BoxDecoration(
                    shape: widget.indicatorShape,
                    color: index == controller.currentPage
                        ? widget.currentIndicatorColor
                        : widget.indicatorColor,
                  ),
                )
              : controller.currentPage == index
                  ? customIndicators[index]['active']
                  : customIndicators[index]['plain'],
        ),
      );
    }

    return SingleChildScrollView(
      scrollDirection: Axis.horizontal,
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: indicators,
      ),
    );
  }
}
